src/Storages/MergeTree/MergeTreeBlockReadUtils.cpp

```diff
@@ -303,7 +303,11 @@ MergeTreeReadTaskColumns getReadTaskColumns(
             if (!columns_from_previous_steps.contains(name))
                 step_column_names.push_back(name);
 
-        if (!step_column_names.empty())
+        bool has_adaptive_granularity = data_part_info_for_reader.getIndexGranularityInfo().mark_type.adaptive;
+
+        /// If part has non-adaptive granularity we always have to read at least one column
+        /// because we cannot determine the correct size of the last granule without reading data.
+        if (!step_column_names.empty() || !has_adaptive_granularity)
```

Comment:
Interesting, what will happen in `injectRequiredColumns()` if `step_column_names` is empty? In theory, execution can get there in case of non-adaptive granularity. Can we cover this case in test?

Reply:
A column with minimal size will be added for reading. That is exactly the case covered by the added test.